source/sessions/snapshot-sessions.rst

```diff
@@ -0,0 +1,353 @@
+
+Test Plan
+=========
+
+Note: some tests are only relevant to certain deployments. For the purpose of deciding
+which tests to run assume that any deployment that is version 4.9 or higher and is either a
+replica set or a sharded cluster supports snapshot reads.
+The server ``minSnapshotHistoryWindowInSeconds`` parameter SHOULD be configured to match the test execution time.
+
+1.  | The first read in a snapshot session must not send
+    | ``atClusterTime`` to the server (because the ``atClusterTime`` has not yet
+    | been determined)
+        * ``session = client.startSession(isSnapshot = true)``
+        * ``document = collection.anyReadOperation(session, ...)``
+        * capture the command sent to the server (using APM or other mechanism)
+        * assert that the command does not have an ``atClusterTime``
+
+2.  | Subsequent snapshot reads on a ``ClientSession`` should read from the snapshot of the first read in that session.
+    | Test SHOULD introduce two variations one for primaries and additional for secondary reads.
+        * ``session1 = client.startSession(isSnapshot = true)``
+        * ``session2 = client.startSession(isSnapshot = true)``
+        * ``readBeforeUpdateSession1 = collection.anyReadOrOperation(session1, ...)``
+        * ``collection.anyUpdateOpertation(...)``
+        * ``readBeforeUpdateSession2 = collection.anyReadOrOperation(session2, ...)``
+        * ``collection.anyUpdateOpertation(...)``
+        * ``readAfterUpdateSession1 = collection.anyReadOrOperation(session1, ...)``
+        * ``readAfterUpdateSession2 = collection.anyReadOrOperation(session2, ...)``
+        * Assert that `readBeforeUpdateSession1` is equivalent to `readAfterUpdateSession1` and 
+        `readBeforeUpdateSession2` to `readAfterUpdateSession2`
+
+3.  | Snapshot reads that fail with ``SnapshotError`` are not retried.
+    | This test is OPTIONAL as it requires setting ``session.snapshotClusterTime`` which is not part of public API.
+        * ``session = client.startSession(isSnapshot = true)``
+        * ``session.snapshotClusterTime = 0``
+        * ``collection.anyReadOrOperation(session1, ...)``
+        * Assert that only single read command was issued
+
+4.  | A read operation in a ``ClientSession`` that is not snapshot read
+    | should not include the ``atClusterTime`` parameter in the command sent to the
+    | server
+        * ``session = client.startSession(isSnapshot = false)``
+        * ``collection.anyReadOperation(session, {})``
+        * ``operationTime = session.operationTime``
+        * capture the command sent to the server (using APM or other mechanism)
+        * assert that the command does not have an ``atClusterTime`` field
+
+Motivation 
+==========
+
+To support snapshot reads. Only supported with server version 4.9 or newer.
+
+Design Rationale
+================
+
+The goal is to modify the driver API as little as possible so that existing
```

Comment:
Let's add a rationale for why we chose this implicit approach rather than the more explicit approach that was suggested in the original scope.

Reply:
Added.